File: bindings/python/examples/simple.py
#!/usr/bin/env python
# simple.py -- basic ppamltracer-python example
# This file is in the public domain.

import os
import sys

from ppamltracer import Tracer

def main():
    # Disable buffering on stdout so we can see the numbers as they are printed.
    sys.stdout = os.fdopen(sys.stdout.fileno(), 'w', 0)
    # Start ppamltracer.
    with Tracer("/tmp/simple_report") as tracer:
        # Register the factorial phase.
        with tracer.create_phase("fact") as phase:
            # Print factorials.
            print "Factorials:",
            for i in range(21):
                print fact(phase, i),
            print
        # Register the Fibonacci phase.
        with tracer.create_phase("fib") as phase:
            # Print Fibonacci numbers.
            print "Fibonacci numbers: ",
            for i in range(24):
                print fib(phase, i),
            print

def fact(phase, n):
    # Record that we're running inside the factorial phase.
    with phase.running():
        # Compute the factorial.
        if n == 0:
            return 1
        else:
            return n * fact(phase, n - 1)

def fib(phase, n):
    # Record that we're running inside the Fibonacci phase.
    with phase.running():
        # Compute the nth Fibonacci number.
        if n == 0 or n == 1:
            return n
        else:
            return fib(phase, n - 1) + fib(phase, n - 2)

if __name__ == '__main__':
    main()

File: bindings/haskell/examples/simple.hs
{- simple.hs -- basic ppamltracer example
This file is in the public domain.

Compile this with
    ghc --make simple.hs
-}

{-# LANGUAGE LambdaCase #-}
module Main where

import Control.Applicative ((<$>), (<*>))
import Control.Monad (liftM)

import PPAML.Tracer

main :: IO ()
main = do
  -- Start ppamltracer.
  withTracer "/tmp/simple_report" $ \tracer -> do
    -- Register the factorial phase.
    withPhase tracer "fact" $ \phase -> do
      -- Print factorials.
      putStr "Factorials:"
      mapM_ (putStr . (' ':) . show) =<< mapM (fact phase) [0 .. 40]
      putStrLn ""
    -- Register the Fibonacci phase.
    withPhase tracer "fib" $ \phase -> do
      -- Print Fibonacci numbers.
      putStr "Fibonacci numbers:"
      mapM_ (putStr . (' ':) . show) =<< mapM (fib phase) [0 .. 25]
      putStrLn ""

{- Records that we're running inside the provided phase and computes a
factorial. -}
fact :: PhaseHandle -> Integer -> IO Integer
fact phase = withPhaseRunning phase . \case
               0 -> return 1
               n -> liftM (n*) $ fact phase (n - 1)

{- Records that we're running inside the provided phase and computes a Fibonacci
number. -}
fib :: PhaseHandle -> Integer -> IO Integer
fib phase = withPhaseRunning phase . \case
              0 -> return 0
              1 -> return 1
              n -> (+) <$> fib phase (n - 1) <*> fib phase (n - 2)
